teps/0094-configuring-resources-at-runtime.md

```diff
@@ -52,12 +61,472 @@ While this may be a valuable feature, it should be considered in a separate prop
+  SidecarOverrides []TaskRunSidecarOverride
+}
+```
+
+### Example Task and TaskRun
+
+Example `Task`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: image-build-task
+spec:
+  steps:
+    - name: build
+      image: gcr.io/kaniko-project/executor:latest
+      command:
+        - /kaniko/executor
+```
+
+Example `TaskRun`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: image-build-taskrun
+spec:
+  taskRef:
+    name: image-build-task
+  stepOverrides:
+    - taskRunStepName: build
+      resources:
+        requests:
+          memory: 1Gi
+```
+
+### Example Pipeline and PipelineRun
+
+Example `Pipeline`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: image-build-pipeline
+spec:
+  tasks:
+    - name: image-build-task
+      steps:
+       - name: build
+         image: gcr.io/kaniko-project/executor:latest
+         command:
+           - /kaniko/executor
+```
+
+Example `PipelineRun`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: image-build-pipelinerun
+spec:
+  taskRunSpecs:
+    - pipelineTaskName: image-build-task
+      stepOverrides:
+        - taskRunStepName: build
+          resources:
+            requests:
+              memory: 1Gi
+```
+
+### Notes/Caveats
+
+#### Mapping container overrides to Steps and Sidecars
+
+This TEP proposes mapping container overrides to their corresponding `Step`s and `Sidecar`s
+using named subobjects, as recommended by
+Kubernetes [API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps).
+This strategy is consistent with other parts of the Tekton API,
+such as the use of
+[`PipelineTaskRunSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#specifying-taskrunspecs)
+to specify `TaskRun` configuration for a `Task` in a `Pipeline`.
+It also meets the requirement that users can specify resource requirements for individual `Step`s and `Sidecar`s.
+
+`Step`s and `Sidecar`s are treated separately because they have different fields.
+We may at some point want to override `Step` fields that are not present in `Sidecar`,
+or vice versa. In addition, a `Step` could share a name with a `Sidecar`;
+separating `StepOverrides` and `SidecarOverrides` avoids ambiguity in this case.
+Names that don't match `Step` or `Sidecar` names will result in the `TaskRun` being rejected.
+
+Some users may want to have resource requirements apply to every `Step` or `Sidecar`.
+This TEP proposes that `TaskRunStepOverride`s without `TaskRunStepName`s specified will apply
+to all `Step`s in the corresponding `Task`, and likewise for `Sidecar`s.
+If such a `TaskRunStepOverride` is present, it must be the only entry in `StepOverrides`.
+This will be the only way to override resource requirements of unnamed `Step`s or `Sidecar`s.
```

Comment:
Could we provide an option to address steps by index instead?

Reply:
I'm confused about this comment in light of your comment below that "the order of steps is not what we consider as the external stable API of a task". What is the reason you would like this option?

Reply:
Yes, sorry, I was adding comments as I was reviewing and forgot to remove this one 🙏